AK/JsonParser.cpp

```diff
@@ -251,7 +251,7 @@ ErrorOr<JsonValue> JsonParser::parse_number()
         }
 
         StringView fraction_string(fraction_buffer.data(), fraction_buffer.size());
-        auto fraction_string_uint = fraction_string.to_uint();
+        auto fraction_string_uint = fraction_string.to_uint<u64>();
         if (!fraction_string_uint.has_value())
             return Error::from_string_literal("JsonParser: Error while parsing number"sv);
         int fraction = fraction_string_uint.value();
```

Comment:
Should `fraction` also be changed to a `u64`? (or just `auto`)

Reply:
I attempted this, but the rounding errors seemed worse when converting both fields to <u64>, this way seemed to have the least rounding error (although it can still be as high as almost 0.5, this absolutely needs to be refactored as a whole.

Reply:
So the problem with the current solution is that if we parse fraction that exceeds 32 bits, we're going to have very wrong results when that fraction is cast from a `u64` to an `int`:

```js
> JSON.parse("1644452550.6489999294281")
1644452550.230759143829345
```

IMO that's actually worse than rejecting a value we knew we couldn't handle.

The rounding errors you encounter aren't really rounding errors, just overflow when values are implicitly cast to `int`. After changing `fraction` to a `u64`, the next issue is that `divider` is also an `int`, and overflows in the `for` loop. We get much more accurate results with the following patch:

```diff
diff --git a/AK/JsonParser.cpp b/AK/JsonParser.cpp
index bf27305938..46f931668b 100644
--- a/AK/JsonParser.cpp
+++ b/AK/JsonParser.cpp
@@ -8,6 +8,7 @@
 #include <AK/JsonArray.h>
 #include <AK/JsonObject.h>
 #include <AK/JsonParser.h>
+#include <math.h>
 
 namespace AK {
 
@@ -251,17 +252,14 @@ ErrorOr<JsonValue> JsonParser::parse_number()
         }
 
         StringView fraction_string(fraction_buffer.data(), fraction_buffer.size());
-        auto fraction_string_uint = fraction_string.to_uint();
+        auto fraction_string_uint = fraction_string.to_uint<u64>();
         if (!fraction_string_uint.has_value())
             return Error::from_string_literal("JsonParser: Error while parsing number"sv);
-        int fraction = fraction_string_uint.value();
-        fraction *= (whole < 0) ? -1 : 1;
+        auto fraction = static_cast<double>(fraction_string_uint.value());
+        double sign = (whole < 0) ? -1 : 1;
 
-        auto divider = 1;
-        for (size_t i = 0; i < fraction_buffer.size(); ++i) {
-            divider *= 10;
-        }
-        value = JsonValue((double)whole + ((double)fraction / divider));
+        auto divider = pow(10.0, static_cast<double>(fraction_buffer.size()));
+        value = JsonValue((double)whole + sign * (fraction / divider));
     } else {
 #endif
         auto to_unsigned_result = number_string.to_uint<u64>();
```

Now we get:
```js
> JSON.parse("1644452550.6489999294281")
1644452550.6489999294281
> JSON.parse("-1644452550.6489999294281")
-1644452550.6489999294281
```

Reply:
BTW we have the same 32-bit issue parsing the integral part of the number (which can of course be fixed separately, just wanted to point it out)

Reply:
Alright, I brought these changes into the PR. Thanks for looking at this as well 👍